### PR TITLE
Add content with solution code and tests

### PR DESCRIPTION
**Pull request by thomastuts**
~~This is not supposed to be merged yet — this lab serves as a proof-of-concept for our testing strategy.~~

### CODE REVIEW

**Comment by thomastuts** [comment]
I overhauled bits of the testing strategy. Tests can now be run in the normal way (`npm test` etc) as well as in the browser (using `mocha.run()`). There are two bundles that get included in the browser:
-  `bundle.js` — the compiled source code of the lab content
- `browser-test-bundle.js` — the compiled source code of the tests, this allows us to run the tests in the browser too. We need to make sure that we run `npm run bundle:test` when we change tests since this file is not compiled on-the-fly.

**Comment by abhishekpillai** [comment]
@thomastuts Looks good overall! Out of curiosity, how can we make sure this happens -> `We need to make sure that we run npm run bundle:test when we change tests since this file is not compiled on-the-fly.`

Possibly, a git hook?

**Comment by thomastuts** [comment]
@abhishekpillai I like the sound of that. I'm not super familiar with Git hooks, but it looks like you can't commit these to a repository. I think we'd need some kind of script that symlinks our hooks into the `.git` repository, right?

**Comment by abhishekpillai** [comment]
@thomastuts I haven't built a pre-commit hook in a while. this post looks like a simple intro: http://johnkpaul.com/blog/2013/10/04/git-precommit-hook-awesomeness/

File: package.json [hunk 1]
```diff
@@ -10,7 +10,11 @@
   "scripts": {
     "debug": "node-debug --hidden node_modules _mocha --watch",
     "debug-ide": "./bin/debug-ide",
-    "test": "mocha -R mocha-multi --reporter-options nyan=-,json=.results.json"
+    "copy-git-hooks": "./bin/copy-hooks",
+    "test": "mocha -R mocha-multi --require test/root.js --reporter-options nyan=-,json=.results.json",
+    "bundle": "browserify index.js -t babelify --exclude react/addons --exclude react/lib/ExecutionEnvironment --exclude react/lib/ReactContext -o bundle.js",
+    "bundle:test": "browserify test/index-test.js -t babelify --exclude react/addons --exclude react/lib/ExecutionEnvironment --exclude react/lib/ReactContext -o browser-test-bundle.js",
```
Comment (pletcher) [inline package.json]: Are these exclusions reliable? The only reason I ask is that it seems like the React team likes to move stuff around -- we'll have to update these lines if something gets out of whack here.

Comment (pletcher) [inline package.json]: Are these exclusions reliable? The only reason I ask is that it seems like the React team likes to move stuff around -- we'll have to update these lines if something gets out of whack here.

**Comment by pletcher** [comment]
@thomastuts, this looks solid -- I noted one question, but I don't think it's a blocker (nor do I think that I am in any position to block this; just wanted to add another :thumbsup: :) )

**Comment by SophieDeBenedetto** [comment]
Went through and tried to complete the lab. Instructions look good, was able to get all tests passing when running `npm test` in terminal and `mocha.run()` in browser. Two things:
1. What kind of instruction around running tests are we going to provide students? Will it be within each lab (i.e. "remember to run `npm test` etc..") or is there going to be some kind of into to testing React labs on Learn reading? 
2. I'm a big fan of the Mad Max reference but unfortunately I don't think its 100% appropriate for the broader Learn audience. Can we keep cars but take out the Fury Road?

@AnnJohn

File: package.json (hunk 1 above)
Comment (thomastuts) [inline package.json]: They seem to be reliable _for now_, but like you mentioned, they like moving things around. When it gets to that point, we can run a script to update the `bundle:test` command — I already wrote a script for updating a set of repositories so we can retool that if it ever gets to that point.

Come to think of it, if they do decide to update it, I think the exclusions won't break anything. From what I've read in the Enzyme issues/docs, it seems like they're going to include this stuff in other npm packages on the top level, so these exclusions shouldn't matter much.

**Comment by thomastuts** [comment]
@SophieDeBenedetto 
1. That's actually a very good question, is there some kind of 'prelude' that we can show students before they start a course? Something like a 'Unit 0' to let them know how things work in the course in general, as well as the labs specifically? (cc @AnnJohn)
2. Is it considered inappropriate due to the nature of the movie, or because the general Learn.co audience doesn't care much for that kind of movie (you guys obviously know the students)? Just asking to make sure I keep an eye on any future content that might be inappropriate.

**Comment by SophieDeBenedetto** [comment]
hey @thomastuts, to answer your second question, i think a little bit of both. That movie is pretty violent and we just can't predict what kind of reaction such a wide range of students will have, also even though it was super popular I think its not quite universally relevant enough to make an understandable example for everyone. Think Doctor Who vs. Star Wars--everyone knows what Star Wars is and gets the reference, even if you haven't seen them, and not so much with Doctor Who (which is better, obviously :) )

**Comment by thomastuts** [comment]
Sure, that makes sense. I'll update the references to something more relevant!

**Comment by ezmiller** [comment]
@thomastuts already a lot of conversation here, so seems as thought this has  been well-reviewed already? It looks pretty solid to me. I think it can be merged. Only odd thing I noticed -- thought it may not be a problem at all -- is that when I ran the tests with mocha, a lot of spaces was generated after some of the tests outputs. Not sure why.

**Comment by thomastuts** [comment]
@ezmiller Interesting, I'm not getting that in my console output. I'll merge this, but let me know if this pops up anywhere else (and if it is affecting tests at all — I suspect it's not).

**Pull request merged**